### ISSUE TITLE
Allow default to be set on the objectType

### ISSUE THREAD

**Issue by Sytten** [Sytten/nexus-shield#50]
It would be nice to be able to set a default rule on the object type with overrides on each field that defines one.

**Comment by Sytten** [comment]
Partially fixed by #59, but it breaks the typing and I still have no idea why.

**Comment by Sytten** [comment]
@lewebsimple Not sure how good you are with advanced typescript but I could use a hand on that one if you have time.
I put most details in there: https://github.com/graphql-nexus/nexus/issues/568
The issue is quite easy to reproduce and I could not find a good fix for it.

**Comment by lewebsimple** [comment]
@Sytten hmmm I'm no TypeScript wizard so I compared your code to the official fieldAuthorizePlugin which should be similar in scope.

At first glance, the corresponding types `FieldShieldResolver` and `FieldAuthorizeResolver` differ in a significant way.. while `FieldAuthorizeResolver` has a (function) signature of `(root, args, ctx, info) => MaybePromise<boolean | Error>`, yours is of type `{ resolve: (root, args, ctx, info, options) => Promise<ShieldRuleResult> }`.

So maybe using the `.resolve` type instead of the (parent) `FieldShieldResolver` could fix that ?

Also, I fail to see the difference between `ShieldRule` and `ShieldRuleFunction`.

**Comment by lewebsimple** [comment]
It also occurs to me that the generated type differ from what you state in the other issue:

This is what you wrote:
```
interface NexusGenPluginTypeConfig<TypeName extends string> {
  shield?: ShieldRule<TypeName, any>
}
```

This is what I get in my project:
```
interface NexusGenPluginTypeConfig<TypeName extends string> {
  shield?: ObjectTypeShieldResolver<TypeName>;
}
```
where `ObjectTypeShieldResolver<TypeName>` corresponds to `ShieldRule<TypeName, never>`

Should it be `any` instead of `never` ?

**Comment by Sytten** [comment]
Yeah I tried to simplify it, the `ObjectTypeShieldResolver` is just an alias type that plugins are supposed to use:
https://github.com/Sytten/nexus-shield/blob/master/src/plugin.ts#L48
Its import process that is a bit strange.
For some reason TS cant infer the real type and falls back to `string` which backs populate to the object and messes up the whole typing.

**Comment by lewebsimple** [comment]
Did you base your code off of the official fieldAuthorize plugin ?

**Comment by Sytten** [comment]
Similar yes, but I reused the class based system of graphql-shield. I did my best on the TS trickery (some of it is explained at the end of the README), but this last problem I really have no clue. I would need better AST debugging tooling to understand why the compiler cant infer the parameter on object but can on the resolvers.

**Comment by talentlessguy** [comment]
does it still not work?

`objectType.shield` seems to not trigger:

```ts
export const Post = objectType({
  name: 'Post',
  definition(t) {
    t.model.author()
    t.model.title()
    t.model.sourceId()
    t.model.content()
    t.model.createdAt()
    t.model.published()
    t.model.authorId()
  },
  shield: {
    resolve: async (root, args, ctx, info, options) => {
      console.log('here')

      return false
    }
  }
})
```

console: empty

**Comment by sooxt98** [comment]
Im getting this error when using `t.field`, `t.string` works fine
![image](https://user-images.githubusercontent.com/13378059/134919421-43b206ac-b843-4c52-b729-7bd302e38efe.png)

**Comment by Sytten** [comment]
I am not really sure why the typing is broken. I don't have a lot of time to work that, so if it is important for you I would suggest you take a look at the code (will review PRs).

**Comment by Sytten** [comment]
Just a note for later, with the new in/out keyword for typescript there is probably a way to force the compiler to understand that the type used is an `in` and it should not override the assignee typing.
